analyses/gene-set-enrichment-analysis/util/hallmark_models.R

```diff
@@ -0,0 +1,112 @@
+
+library(dplyr)
+library(purrr)
+library(tidyr)
+library(stringr)
+library(broom)
+
+
+# Magrittr pipe
+`%>%` <- dplyr::`%>%`
+
+### Check tidyr version for nesting code to ensure it used as <1.0.0 ####
+tidyr_version <- package_version( packageVersion("tidyr") )
+if (tidyr_version$major >= 1){
+  nest   <- nest_legacy
+  unnest <- unnest_legacy
+}
+
+
+
+
+
+### Define function for performing ANOVA, and subsequently posthoc Tukey tests, for identifying pathways significantly different across histologies, and significant pairwise comparisons of histologies within a pathway. 
+
+
+
+perform_anova <- function(df, predictor_variable_string)
+{
+  ### Defines a function for performing ANOVA to detect signficance GSVA scores according to a given predictor variable
+  ### Specifically to be used within `compare_pathways()`
+  formula <- as.formula(paste("gsea_score ~", predictor_variable_string))
+  aov(formula, data = df)
+}
+
+gsva_anova_tukey <- function(df, predictor_variable, library_type, significance_threshold)
+{
+  ### Defines a function to compare GSVA scores across a given predictor variable, _for each pathway (hallmark)_.
+  ### Returns a named list with two items:
+  ##### 1. "anova" is a tibble of all ANOVAs performed, for each hallmark
+  ##### 2. "tukey" is a tibble of all Tukey HSD tests performed, comparing each level of `predictor_variable` for each hallmark
+  
+  ### Input arguments:
+  ### 1. `df`:   the data frame containing scores (`gsea_score`) and assumed to contain the specified `predictor_variable` 
+  ### 2. `predictor_variable`: A _string_ indicating the predictor variable, such as `short_histology`, `disease_type_new`, etc.
+  ### 3. `library_type`:  Which RNAseq library to analyze? Specify either `polyA` or `stranded`
+  ### 4. `significance_threshold`:  What threshold is used for significance (aka alpha, FPR)?
+
+  predictor_variable        <- dplyr::enquo(predictor_variable)  
+  predictor_variable_string <- quo_name(predictor_variable)
+  library_type <- str_to_lower(library_type)
+  ######## Sanity check #########
+  if (!(predictor_variable_string %in% colnames(df))) stop("ERROR: The provided `predictor_variable` is not found in the data frame.")
+  if (!("hallmark_name" %in% colnames(df))) stop("ERROR: The needed column `hallmark_name` is not found in the data frame.")
+  if (!("gsea_score" %in% colnames(df))) stop("ERROR: The needed column `gsea_score` is not found in the data frame.")
+  if (!(library_type %in% unique(df$data_type))) stop(paste("ERROR: The provided `library_type` must be one of:", 
+                                                         paste(unique(df$data_type), collapse= " or "),
+                                                         "."))
+
+  ### Determine how many ANOVAs we will be doing
+  df %>% 
+    dplyr::select(hallmark_name, data_type) %>%
+    distinct() %>% 
+    count(data_type) %>%
```

Comment:
It's unreasonably confusing, will streamline.

Reply:
Strike that, this is the way to do it. Count will not update after filtering, but the filter is needed in case there are different numbers of pathways for each library.